In [7]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report, recall_score

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, VotingClassifier

# Creating a Baseline Model

In [4]:
df = pd.read_csv('novagen_dataset.csv')
df.head()

,Age,BMI,Blood_Pressure,Cholesterol,Glucose_Level,Heart_Rate,Sleep_Hours,Exercise_Hours,Water_Intake,Stress_Level,...,Diet,MentalHealth,PhysicalActivity,MedicalHistory,Allergies,Diet_Type__Vegan,Diet_Type__Vegetarian,Blood_Group_AB,Blood_Group_B,Blood_Group_O
0,2.0,26.0,111.0,198.0,99.0,72.0,4.0,1.0,5.0,5.0,...,1,2,1,0,1,False,True,True,False,False
1,8.0,24.0,121.0,199.0,103.0,75.0,2.0,1.0,2.0,9.0,...,1,2,1,2,2,False,False,True,False,False
2,81.0,27.0,147.0,203.0,100.0,74.0,10.0,-0.0,5.0,1.0,...,2,0,0,1,0,True,False,False,False,False
3,25.0,21.0,150.0,199.0,102.0,70.0,7.0,3.0,3.0,3.0,...,1,2,1,2,0,True,False,False,True,False
4,24.0,26.0,146.0,202.0,99.0,76.0,10.0,2.0,5.0,1.0,...,2,0,2,0,2,False,True,False,True,False


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9549 entries, 0 to 9548
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Age                    9549 non-null   float64
 1   BMI                    9549 non-null   float64
 2   Blood_Pressure         9549 non-null   float64
 3   Cholesterol            9549 non-null   float64
 4   Glucose_Level          9549 non-null   float64
 5   Heart_Rate             9549 non-null   float64
 6   Sleep_Hours            9549 non-null   float64
 7   Exercise_Hours         9549 non-null   float64
 8   Water_Intake           9549 non-null   float64
 9   Stress_Level           9549 non-null   float64
 10  Target                 9549 non-null   int64  
 11  Smoking                9549 non-null   int64  
 12  Alcohol                9549 non-null   int64  
 13  Diet                   9549 non-null   int64  
 14  MentalHealth           9549 non-null   int64  
 15  Phys

In [6]:
df.shape

(9549, 23)

In [8]:
# Split features and target
X = df.drop(columns=["Target"])
y = df["Target"]

In [9]:
# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y)

In [11]:
ss = StandardScaler()

x_train_s = ss.fit_transform(X_train)
x_test_s = ss.transform(X_test)

In [23]:
# Logistic Regression (with Regularization)

lr = LogisticRegression(penalty='l2', max_iter=1000)

lr.fit(x_train_s,y_train)

y_pred_lr = lr.predict(x_test_s)
print('Logictic Regression Accuracy:',accuracy_score(y_test,y_pred_lr))
print('Logictic Regression Recall:', recall_score(y_test, y_pred_lr))
print(classification_report(y_test,y_pred_lr))

Logictic Regression Accuracy: 0.8136125654450261
Logictic Regression Recall: 0.8273092369477911
              precision    recall  f1-score   support

           0       0.81      0.80      0.80       914
           1       0.82      0.83      0.82       996

    accuracy                           0.81      1910
   macro avg       0.81      0.81      0.81      1910
weighted avg       0.81      0.81      0.81      1910



# Model 2 - KNN

In [22]:
knn = KNeighborsClassifier(n_neighbors=5, metric='euclidean')

knn.fit(x_train_s,y_train)

y_pred_knn = knn.predict(x_test_s)
print('KNN Accuracy:',accuracy_score(y_test,y_pred_knn))
print('KNN Recall:', recall_score(y_test, y_pred_knn))
print(classification_report(y_test,y_pred_knn))

KNN Accuracy: 0.8832460732984293
KNN Recall: 0.8835341365461847
              precision    recall  f1-score   support

           0       0.87      0.88      0.88       914
           1       0.89      0.88      0.89       996

    accuracy                           0.88      1910
   macro avg       0.88      0.88      0.88      1910
weighted avg       0.88      0.88      0.88      1910



# Model 3 - Random Forest (Ensemble Learning)

In [18]:
rfc = RandomForestClassifier(n_estimators=200,random_state=42)

rfc.fit(x_train_s,y_train)

y_pred_rfc = rfc.predict(x_test_s)
print('Random Forest Accuracy:',accuracy_score(y_test,y_pred_rfc))
print('Random Forest Recall:', recall_score(y_test, y_pred_rfc))
print(classification_report(y_test,y_pred_rfc))

lr accuracy_score: 0.9397905759162304
lr Recall score: 0.9618473895582329
classification report:               precision    recall  f1-score   support

           0       0.96      0.92      0.94       914
           1       0.93      0.96      0.94       996

    accuracy                           0.94      1910
   macro avg       0.94      0.94      0.94      1910
weighted avg       0.94      0.94      0.94      1910



# Gradient Boosting (Ensemble Learning)

In [24]:
gbc = GradientBoostingClassifier(n_estimators=200,
                                 learning_rate=0.1,
                                max_depth=3,
                                random_state=42)

gbc.fit(x_train_s,y_train)

y_pred_gbc = gbc.predict(x_test_s)
print('Gradient Boosting Accuracy:',accuracy_score(y_test,y_pred_gbc))
print('Gradient Boosting Recall:', recall_score(y_test, y_pred_gbc))
print(classification_report(y_test,y_pred_gbc))

Gradient Boosting Accuracy: 0.9350785340314136
Gradient Boosting Recall: 0.9528112449799196
              precision    recall  f1-score   support

           0       0.95      0.92      0.93       914
           1       0.92      0.95      0.94       996

    accuracy                           0.94      1910
   macro avg       0.94      0.93      0.93      1910
weighted avg       0.94      0.94      0.94      1910



# Voting Classifier (Ensemble Learning)

In [25]:
vc = VotingClassifier(
    estimators=[
        ('lr',lr),
        ('knn',knn),
        ('rfc',rfc)],
    voting='soft')

vc.fit(x_train_s,y_train)

y_pred_vc = vc.predict(x_test_s)
print('Voting Classifier Accuracy:',accuracy_score(y_test,y_pred_vc))
print('Voting Classifier Recall:', recall_score(y_test, y_pred_vc))
print(classification_report(y_test,y_pred_vc))

Voting Classifier Accuracy: 0.9157068062827225
Voting Classifier Recall: 0.929718875502008
              precision    recall  f1-score   support

           0       0.92      0.90      0.91       914
           1       0.91      0.93      0.92       996

    accuracy                           0.92      1910
   macro avg       0.92      0.92      0.92      1910
weighted avg       0.92      0.92      0.92      1910



## Results

| Model                | Recall |
|----------------------|:------:|
| Logistic Regression  | 82.7%  |
| KNN                  | 88.4%  |
| Random Forest        | 96.2%  |
| Gradient Boosting    | 95.3%  |
| Voting Classifier    | 93.0%  |

### Best Classifier that we should use for NovaGen(based on Recall) - Random Forest with accuracy of 96.2%